lane.py

```diff
@@ -1306,364 +1215,127 @@ def search(self, query, search_client, pagination=None):
-                    )
-                    q = self.only_show_ready_deliverable_works(q, mw)
-                    q = self._defer_unused_opds_entry(q, work_model=mw)
-                    work_by_id = dict()
-                    a = time.time()
-                    works = q.all()
-                    for mw in works:
-                        work_by_id[mw.works_id] = mw
-                    results = [work_by_id[x] for x in doc_ids if x in work_by_id]
-                    b = time.time()
-                    logging.debug(
-                        "Obtained %d MaterializedWork objects in %.2fsec",
-                        len(results), b-a
-                    )
-
-        if not results:
-            logging.debug("No elasticsearch results, falling back to database query")
-            results = self._search_database(query).limit(pagination.size).offset(pagination.offset).all()
+                    results = self.works_for_specific_ids(_db, doc_ids)
+
         return results
 
-    def _search_database(self, query):
-        """Do a really awful database search for a book using ILIKE.
+    def apply_custom_filters(self, _db, qu, work_model, featured=False):
+        """Apply filters to a base query against a materialized view,
+        yielding a query that only finds books in this Lane.
 
-        This is useful if an app server has no external search
-        interface defined, or if the search interface isn't working
-        for some reason.
-        """
-        k = "%" + query + "%"
-        q = self.works().filter(
-            or_(Edition.title.ilike(k),
-                Edition.author.ilike(k)))
-        #q = q.order_by(Work.quality.desc())
-        return q
-
-    def _defer_unused_opds_entry(self, query, work_model=Work):
-        """Defer the appropriate opds entry
+        :param work_model: Either MaterializedWork or MaterializedWorkWithGenre
         """
-        if Configuration.DEFAULT_OPDS_FORMAT == "simple_opds_entry":
-            return query.options(defer(work_model.verbose_opds_entry))
-        else:
-            return query.options(defer(work_model.simple_opds_entry))
-
-    def sublane_samples(self, use_materialized_works=True):
-        """Generates a list of samples from each sublane for a groups feed"""
-
-        # This is a list rather than a dict because we want to
-        # preserve the ordering of the lanes.
-        works_and_lanes = []
-        for sublane in self.visible_sublanes:
-            works = sublane.featured_works(
-                use_materialized_works=use_materialized_works
+        parent_distinct = False
+        if self.parent and self.inherit_parent_restrictions:
+            # In addition to the other restrictions imposed by this
+            # Lane, books will show up here only if they would
+            # also show up in the parent Lane.
+            qu, parent_distinct = self.parent.apply_bibliographic_filters(
+                _db, qu, work_model, featured
```

Comment:
I think I understand what this is doing, but the relationship between `apply_bibliographic_filters` and `apply_custom_filters` is confusing. `apply_bibliographic_filters` calls `apply_custom_filters`, but then `apply_custom_filters` calls `apply_bibliographic_filters` on the parent lane. I'm not sure why there are two methods.

Reply:
The subclasses of `WorkList` in the circulation manager, like `SeriesLane`, need to override `apply_custom_filters` but leave `apply_bibliographic_filters` alone. I think I can replace this with one method where the subclasses call `super()` though.